src/sp/it/pl/audio/playlist/PlaylistReader.kt

```diff
@@ -0,0 +1,91 @@
+package sp.it.pl.audio.playlist
+
+import mu.KotlinLogging
+import sp.it.pl.audio.Item
+import sp.it.pl.audio.SimpleItem
+import sp.it.pl.util.dev.fail
+import sp.it.pl.util.dev.throwIfFxThread
+import sp.it.pl.util.file.div
+import sp.it.pl.util.file.hasExtension
+import sp.it.pl.util.file.parentDir
+import sp.it.pl.util.functional.net
+import java.io.File
+import java.io.IOException
+import java.net.URI
+import java.net.URISyntaxException
+import java.nio.charset.Charset
+
+private const val EXTM3U = "#EXTM3U"
+private const val EXTINF = "#EXTINF"
+private val logger = KotlinLogging.logger { }
+
+@Suppress("ASSIGNED_BUT_NEVER_ACCESSED_VARIABLE")
+fun readPlaylist(file: File): List<Item> {
+    throwIfFxThread()
+
+    val location = file.parentDir!!
+    val encoding = when {
+        file hasExtension "m3u" -> Charset.defaultCharset()
+        file hasExtension "m3u8" -> Charsets.UTF_8
+        else -> fail { "File=$file is not a supported playlist file" }
+    }
+
+    file.useLines(encoding) { lines ->
+        var isFirstLine = true
+        var isExtended = false
+        return lines
+                .onEach { if (isFirstLine && EXTM3U==it) isExtended = true }
+                .onEach { isFirstLine = false }
+                .filter { !it.startsWith("#") && !it.isEmpty() }
+                .flatMap {
+                    null
+                        ?: it.toURIOrNull()?.net { sequenceOf(SimpleItem(it)) }
+                        ?: File(it).absoluteTo(location)
+                                ?.net {
+                                    if (it.isPlaylistFile()) readPlaylist(it).asSequence()
+                                    else sequenceOf(SimpleItem(it))
+                                }
+                                ?: sequenceOf()
+                }
+                .toList()
+    }
+}
+
+fun writePlaylist(playlist: List<Item>, name: String, location: File) {
+    throwIfFxThread()
```

Comment:
Thread checking is indeed a pain and I have thought about it a lot.

I will add javadoc, but it is not enough.

I thought we already learned from annotation processors... I would like to have none in the project. But your idea is very neat, I give you that. As a last resort, maybe.

I entertained an idea of 'propagating' annotation like NotNull that would issue a warning if method with it is called on javaFx thread (this would probably require annotating all javafx code so no deal). Then I thought about how such method could only ever be called inside a method annotated with another annotation. Basically exactly what coroutines do - suspending function can only be called inside coroutine. So we could in theory refactor all async code to suspending functions but I'm kinda not ok with that at this point (if we can get Future to consume these methods natively then hell yes im in) or alternatively, we make all async methods return Fut which will force dev to call it in future chains (where the Fut signature will vanish, hence retain 100% readability on consumer side), but I dislike having to return Fut from functions that we normally just want to mark some way.

Anyway,  out of scope. We will have to agree to tackle this growing issue in the future. But I'm glad you pointed this out. This is a major issue.

Reply:
This will most probably remain a problem though, as any blocking code is a no-go on fx thread and would have to be designated for async only. This is bad, because 1 even stuff like File.isFile() would be affacted, 2 we can not contorl libs and jdk code, 3 this only makes sense from point of view of fx thread, but on other threads we do not want to run these things asynchronously.

So we are stuck
- either dev on javafx thread is prone to block threads with methods he shouldnt be allowed to call
- or we complicate/deoptimize code running on non javafx threads.

Reply:
Good news: It is possible to make it so Fut's then() methods (and the like) consume suspending functions (and run them on the specified executor inside runBlocking). There may be some (probably small) overhead.

Unfortunately calling suspending functions from Java is a no go. So we are blocked. Will be done in the future.

Reply:
I used Blocks annotation. For now good enough.